zero/src/bin/trie_diff.rs

```diff
@@ -0,0 +1,159 @@
+use std::io::Read;
+use std::iter::repeat;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use anyhow::Result;
+use clap::{Parser, ValueHint};
+use evm_arithmetization::generation::DebugOutputTries;
+use futures::{future, TryStreamExt};
+use paladin::directive::{Directive, IndexedStream};
+use paladin::runtime::Runtime;
+use regex::Regex;
+use trace_decoder::observer::TriesObserver;
+use tracing::{error, info};
+use zero::ops::register;
+use zero::prover::{cli::CliProverConfig, BlockProverInput, ProverConfig};
+
+#[derive(Parser)]
+#[command(version = zero::version(), propagate_version = true)]
+pub(crate) struct Cli {
+    #[clap(flatten)]
+    pub(crate) prover_config: CliProverConfig,
+
+    /// The previous proof output.
+    #[arg(long, short = 'f', value_hint = ValueHint::FilePath)]
+    previous_proof: Option<PathBuf>,
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    zero::tracing::init();
+
+    let args = Cli::parse();
+
+    // Load witness input from stdin
+    let mut buffer = String::new();
+    std::io::stdin().read_to_string(&mut buffer)?;
+
+    // Debug run, we always use in-memory execution
+    let paladin_config = paladin::config::Config {
+        amqp_uri: None,
+        runtime: paladin::config::Runtime::InMemory,
+        ..Default::default()
+    };
+    let runtime = Arc::new(Runtime::from_config(&paladin_config, register()).await?);
+
+    // Tries are computed in the kernel so no need to run proving, test_only mode is
+    // enough. We hardcode prover arguments that we need for trie diff computation.
+    let prover_config: Arc<ProverConfig> = Arc::new(ProverConfig {
+        test_only: true,
+        save_inputs_on_error: true,
+        save_tries_on_error: true,
+        ..args.prover_config.into()
+    });
+
+    let seg_ops = zero::ops::SegmentProofTestOnly {
+        save_inputs_on_error: prover_config.save_inputs_on_error,
+        save_tries_on_error: prover_config.save_tries_on_error,
+    };
+
+    let des = &mut serde_json::Deserializer::from_str(&buffer);
+    let block_prover_inputs = serde_path_to_error::deserialize::<_, Vec<BlockProverInput>>(des)?
+        .into_iter()
+        .collect::<Vec<_>>();
+
+    for block_prover_input in block_prover_inputs {
+        let mut observer = TriesObserver::new();
+        let block_number = block_prover_input
+            .other_data
+            .b_data
+            .b_meta
+            .block_number
+            .as_u64();
```

Comment:
I don't like this call because it can panic, but doesn't look like it

Reply:
Yeah these method names aren't great. `low_u64` does the same here without panicking